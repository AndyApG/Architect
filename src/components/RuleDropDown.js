/* eslint-disable jsx-a11y/label-has-for */

import React from 'react';
import { isArray, find, isEmpty } from 'lodash';
import PropTypes from 'prop-types';
import cx from 'classnames';

const withLabels = options =>
  options.map(option => (isArray(option) ? option : [option, option]));

const getSpacer = (optionsWithLabels, placeholder, value) => (
  isEmpty(value) ?
    placeholder : // return placeholder text
    find(optionsWithLabels, [0, value])[1] // return matching label
);

const RuleDropDown = ({
  options,
  onChange,
  placeholder,
  className,
  value,
}) => {
  if (options.length === 0) { return null; }

  const optionsWithLabels = withLabels(options);
  const spacer = getSpacer(optionsWithLabels, placeholder, value);

  return (
    <label className={cx('rule-drop-down', className)} >
      <div className="rule-drop-down__spacer">{ spacer }</div>
      <select
        defaultValue=""
        value={value}
        onChange={e => onChange(e.target.value)}
        className="rule-drop-down__options"
      >
        { placeholder && <option value="" disabled>{placeholder}</option> }
        {optionsWithLabels.map(
          ([optionValue, optionLabel], index) => {
            return <option key={index} value={optionValue}>{optionLabel}</option>;
          },
        )}
      </select>
    </label>
  );
};

RuleDropDown.propTypes = {
  options: PropTypes.array,
  value: PropTypes.string,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  className: PropTypes.string,
};

RuleDropDown.defaultProps = {
  placeholder: '',
  className: null,
  value: '',
  onChange: () => {},
  options: [],
};

export default RuleDropDown;
