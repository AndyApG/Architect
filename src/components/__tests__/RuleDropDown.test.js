/* eslint-env jest */

import React from 'react';
import { toPairs } from 'lodash';
import { shallow } from 'enzyme';
import RuleDropDown from '../RuleDropDown';

const mockProps = {
  options: toPairs({
    foo: 'bar',
    baz: 'buzz',
  }),
  value: '',
  onChange: () => {},
};

describe('<RuleDropDown />', () => {
  it('can render without labels', () => {
    const component = shallow(<RuleDropDown {...mockProps} />);

    expect(component).toMatchSnapshot();

    const firstOption = component.find('option').at(0);
    expect([
      firstOption.prop('value'),
      firstOption.text()
    ]).toEqual(['foo', 'bar']);
  });

  it('can render with labels', () => {
    const options = ['foo', 'bar', 'baz'];
    const component = shallow(<RuleDropDown {...mockProps} options={options} />);

    expect(component).toMatchSnapshot();

    const firstOption = component.find('option').at(0);

    expect([
      firstOption.prop('value'),
      firstOption.text()
    ]).toEqual(['foo', 'foo']);
  });

  it('it triggers on change with the option value', () => {
    const onChange = jest.fn();
    const component = shallow(<RuleDropDown {...mockProps} onChange={onChange} />);

    component.find('select').simulate('change', { target: { value: 'foo' } });
    expect(onChange.mock.calls).toEqual([['foo']]);
  });
});
