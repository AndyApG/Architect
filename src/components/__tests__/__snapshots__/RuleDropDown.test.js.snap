// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RuleDropDown /> can render with labels 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RuleDropDown
    className={null}
    onChange={[Function]}
    options={
        Array [
            "foo",
            "bar",
            "baz",
          ]
    }
    placeholder=""
    value=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="rule-drop-down__spacer"
/>,
        <select
          className="rule-drop-down__options"
          defaultValue=""
          onChange={[Function]}
          value=""
>
          
          <option
                    value="foo"
          >
                    foo
          </option>
          <option
                    value="bar"
          >
                    bar
          </option>
          <option
                    value="baz"
          >
                    baz
          </option>
</select>,
      ],
      "className": "rule-drop-down",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": "rule-drop-down__spacer",
        },
        "ref": null,
        "rendered": "",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "",
            Array [
              <option
                value="foo"
>
                foo
</option>,
              <option
                value="bar"
>
                bar
</option>,
              <option
                value="baz"
>
                baz
</option>,
            ],
          ],
          "className": "rule-drop-down__options",
          "defaultValue": "",
          "onChange": [Function],
          "value": "",
        },
        "ref": null,
        "rendered": Array [
          "",
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": "foo",
              "value": "foo",
            },
            "ref": null,
            "rendered": "foo",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": "bar",
              "value": "bar",
            },
            "ref": null,
            "rendered": "bar",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "host",
            "props": Object {
              "children": "baz",
              "value": "baz",
            },
            "ref": null,
            "rendered": "baz",
            "type": "option",
          },
        ],
        "type": "select",
      },
    ],
    "type": "label",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="rule-drop-down__spacer"
/>,
          <select
            className="rule-drop-down__options"
            defaultValue=""
            onChange={[Function]}
            value=""
>
            
            <option
                        value="foo"
            >
                        foo
            </option>
            <option
                        value="bar"
            >
                        bar
            </option>
            <option
                        value="baz"
            >
                        baz
            </option>
</select>,
        ],
        "className": "rule-drop-down",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": "rule-drop-down__spacer",
          },
          "ref": null,
          "rendered": "",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "",
              Array [
                <option
                  value="foo"
>
                  foo
</option>,
                <option
                  value="bar"
>
                  bar
</option>,
                <option
                  value="baz"
>
                  baz
</option>,
              ],
            ],
            "className": "rule-drop-down__options",
            "defaultValue": "",
            "onChange": [Function],
            "value": "",
          },
          "ref": null,
          "rendered": Array [
            "",
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": "foo",
                "value": "foo",
              },
              "ref": null,
              "rendered": "foo",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": "bar",
                "value": "bar",
              },
              "ref": null,
              "rendered": "bar",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "host",
              "props": Object {
                "children": "baz",
                "value": "baz",
              },
              "ref": null,
              "rendered": "baz",
              "type": "option",
            },
          ],
          "type": "select",
        },
      ],
      "type": "label",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<RuleDropDown /> can render without labels 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RuleDropDown
    className={null}
    onChange={[Function]}
    options={
        Array [
            Array [
              "foo",
              "bar",
            ],
            Array [
              "baz",
              "buzz",
            ],
          ]
    }
    placeholder=""
    value=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="rule-drop-down__spacer"
/>,
        <select
          className="rule-drop-down__options"
          defaultValue=""
          onChange={[Function]}
          value=""
>
          
          <option
                    value="foo"
          >
                    bar
          </option>
          <option
                    value="baz"
          >
                    buzz
          </option>
</select>,
      ],
      "className": "rule-drop-down",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": "rule-drop-down__spacer",
        },
        "ref": null,
        "rendered": "",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "",
            Array [
              <option
                value="foo"
>
                bar
</option>,
              <option
                value="baz"
>
                buzz
</option>,
            ],
          ],
          "className": "rule-drop-down__options",
          "defaultValue": "",
          "onChange": [Function],
          "value": "",
        },
        "ref": null,
        "rendered": Array [
          "",
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": "bar",
              "value": "foo",
            },
            "ref": null,
            "rendered": "bar",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": "buzz",
              "value": "baz",
            },
            "ref": null,
            "rendered": "buzz",
            "type": "option",
          },
        ],
        "type": "select",
      },
    ],
    "type": "label",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="rule-drop-down__spacer"
/>,
          <select
            className="rule-drop-down__options"
            defaultValue=""
            onChange={[Function]}
            value=""
>
            
            <option
                        value="foo"
            >
                        bar
            </option>
            <option
                        value="baz"
            >
                        buzz
            </option>
</select>,
        ],
        "className": "rule-drop-down",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": "rule-drop-down__spacer",
          },
          "ref": null,
          "rendered": "",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "",
              Array [
                <option
                  value="foo"
>
                  bar
</option>,
                <option
                  value="baz"
>
                  buzz
</option>,
              ],
            ],
            "className": "rule-drop-down__options",
            "defaultValue": "",
            "onChange": [Function],
            "value": "",
          },
          "ref": null,
          "rendered": Array [
            "",
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": "bar",
                "value": "foo",
              },
              "ref": null,
              "rendered": "bar",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": "buzz",
                "value": "baz",
              },
              "ref": null,
              "rendered": "buzz",
              "type": "option",
            },
          ],
          "type": "select",
        },
      ],
      "type": "label",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
